<a href="https://colab.research.google.com/github/panchosepq19/football-match-predictor/blob/main/predictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
matches = pd.read_csv("matches.csv", index_col=0)
matches.shape

(1389, 27)

38 matches played oer season and 20 teams in the league. I have the 2021 and 2022 season data. according to the shape, I have 1389 matches when there should be 1520. There is some missing data.


Because of relegation there would be 6 teams with less than 70 matches. However the data shows 7 teams with less than 70 matches.

In [ ]:
matches["team"].value_counts()

team
Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: count, dtype: int64



as seen below the missing data is due to the fact that some teams are only reporting one season in the data whereas other teams are presenting 2 seasons.



In [ ]:
matches[matches["team"] == "Liverpool"]

date   time            comp         round  day venue result   gf  \
1   2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13  2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14  2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16  2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18  2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22  2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23  2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24  2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25  2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27  2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28  2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30  2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31  2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32  2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33  2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34  2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36  2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37  2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38  2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41  2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42  2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44  2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46  2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47  2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk  \
1   3.0     Leeds United  ...  Match Report    NaN  20.0   4.0  17.0  0.0   
2   0.0          Chelsea  ...  Match Report    NaN  17.0   5.0  17.7  1.0   
4   1.0          Arsenal  ...  Match Report    NaN  21.0   9.0  16.8  0.0   
6   7.0      Aston Villa  ...  Match Report    NaN  14.0   8.0  15.8  1.0   
7   2.0          Everton  ...  Match Report    NaN  22.0   8.0  15.0  1.0   
9   1.0    Sheffield Utd  ...  Match Report    NaN  17.0   5.0  18.2  1.0   
11  1.0         West Ham  ...  Match Report    NaN   8.0   2.0  18.6  1.0   
13  1.0  Manchester City  ...  Match Report    NaN   9.0   2.0  21.5  0.0   
14  0.0   Leicester City  ...  Match Report    NaN  24.0  12.0  11.9  0.0   
16  1.0         Brighton  ...  Match Report    NaN   6.0   2.0  20.9  0.0   
18  0.0           Wolves  ...  Match Report    NaN  11.0   6.0  16.6  1.0   
20  1.0           Fulham  ...  Match Report

In [ ]:
matches["round"].value_counts()

round
Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: count, dtype: int64

As seen above we also have some missing data in the 2021-2022 season due to the fact that when the data was collected the season was still going on

##Cleaning Data


In [ ]:
matches["date"] = pd.to_datetime(matches["date"])
matches["hour"] = matches["time"].str.replace(":.+", "", regex=True).astype("int")

In [ ]:
#turning "venue", "opponent" and "date" into a numeric value so I can use it as a prediction input
#(Each oponent will get their own code that can be used in prediction algorithms)
matches["venue_code"] = matches["venue"].astype("category").cat.codes
matches["opp_code"] = matches["opponent"].astype("category").cat.codes
matches["day_code"] = matches["date"].dt.dayofweek

In [ ]:
# target = weather the team won or not
# win = 1, loss and draw = 0
matches["target"] = (matches["result"] == "W").astype("int")

## Training the Model

I am using the 2020-2021 season to predict the 2021-2022 season (whose correct outcomes are also in the data)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [ ]:
rf = RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)
train = matches[matches["date"] < '2022-01-01']
test = matches[matches["date"] > '2022-01-01']

predictors = ["venue_code", "opp_code", "hour", "day_code"]
rf.fit(train[predictors], train["target"])


RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [ ]:
preds = rf.predict(test[predictors])
acc = accuracy_score(test["target"], preds)
prec = precision_score(test["target"], preds)
print(f"The prediction accuracy is {acc*100:.2f}%")
print(f"precision score is {prec*100:.2f}%")

The prediction accuracy is 61.23%
precision score is 47.46%


In [ ]:
combined = pd.DataFrame(dict(actual=test["target"], prediction=preds))
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           141  31
1            76  28

##Improving accuracy

In [ ]:
matches_grouped = matches.groupby("team")
group = matches_grouped.get_group("Manchester City").sort_values("date")

In [ ]:
def rolling_averages(group, cols, new_cols):
  group = group.sort_values("date")
  rolling_stats = group[cols].rolling(3, closed='left').mean()
  group[new_cols] = rolling_stats
  group = group.dropna(subset=new_cols)
  return group

In [ ]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]
rolling_averages(group, cols, new_cols)
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))
matches_rolling = matches_rolling.droplevel('team') # drop the extra index level on the left side of the data set
matches_rolling.index = range(matches_rolling.shape[0]) #give unique indeces to each row

In [ ]:
def make_predictions(data, predictors):
  train = data[data["date"] < '2022-01-01']
  test = data[data["date"] > '2022-01-01']
  rf.fit(train[predictors], train["target"])
  preds = rf.predict(test[predictors])
  combined = pd.DataFrame(dict(actual=test["target"], predicted=preds), index=test.index)
  prec = precision_score(test["target"], preds)
  return combined, prec

In [ ]:
combined, error = make_predictions(matches_rolling, predictors + new_cols)
print(f"precision score is {error*100:.2f}%")


precision score is 62.50%


In [ ]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index=True, right_index=True)


In [110]:
class MissingDict(dict):
  __missing__ = lambda self, key: key

map_values = {"Brighton and Hove Albion": "Brighton",
              "Manchester United": "Manchester Utd",
              "Newcastle United": "Newcastle Utd",
              "Tottenham Hotspur": "Tottenham",
              "West Ham United": "West Ham",
              "Wolverhampton Wanderers": "Wolves"
}

mapping = MissingDict(**map_values)
combined["new_team"] = combined["team"].map(mapping)
combined.shape
merged = combined.merge(combined, left_on=["date", "new_team"], right_on=["date", "opponent"])
merged

actual_x  predicted_x       date                   team_x     opponent_x  \
0           0            0 2022-01-23                  Arsenal        Burnley   
1           1            0 2022-02-10                  Arsenal         Wolves   
2           1            0 2022-02-19                  Arsenal      Brentford   
3           1            1 2022-02-24                  Arsenal         Wolves   
4           1            1 2022-03-06                  Arsenal        Watford   
..        ...          ...        ...                      ...            ...   
257         1            0 2022-03-13  Wolverhampton Wanderers        Everton   
258         0            0 2022-03-18  Wolverhampton Wanderers   Leeds United   
259         1            0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
260         0            0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
261         0            0 2022-04-24  Wolverhampton Wanderers        Burnley   

    result_x new_team_x  actual_y  predicted_y                   team_y  \
0          D    Arsenal         0            0                  Burnley   
1          W    Arsenal         0            0  Wolverhampton Wanderers   
2          W    Arsenal         0            0                Brentford   
3          W    Arsenal         0            0  Wolverhampton Wanderers   
4          W    Arsenal         0            0                  Watford   
..       ...        ...       ...          ...                      ...   
257        W     Wolves         0            0                  Everton   
258        L     Wolves         1            0             Leeds United   
259        W     Wolves         0            0              Aston Villa   
260        L     Wolves         1            0         Newcastle United   
261        L     Wolves         1            0                  Burnley   

    opponent_y result_y     new_team_y  
0      Arsenal        D        Burnley  
1      Arsenal        L         Wolves  
2      Arsenal        L      Brentford  
3      Arsenal        L         Wolves  
4      Arsenal        L        Watford  
..         ...      ...            ...  
257     Wolves        L        Everton  
258     Wolves        W   Leeds United  
259     Wolves        L    Aston Villa  
260     Wolves        W  Newcastle Utd  
261     Wolves        W        Burnley  

[262 rows x 13 columns]

In [111]:
merged[(merged["predicted_x"] == 1) & (merged["predicted_y"] == 0)]["actual_x"].value_counts()
print(f"precision score is {(27/40)*100:.2f}%")

precision score is 67.50%
